[FEATURE] Status update

scribblecrumb opened the pull request:
<!-- Write BELOW The Headers and ABOVE The comments else it may not be viewable. -->
<!-- You can view CONTRIBUTING.md for a detailed description of the pull request process. -->
<!-- Be sure to name your PR something descriptive and succinct; include Bugfix: Feature: Enhancement: or Content: in the title to describe what type of PR it is. -->
<!-- IF YOU ARE DOING A BUGFIX: Please target the latest release branch if the bug that you are fixing is also present in the latest release. -->

## About The Pull Request
It's a doozy.  This completely reworks our status system!

`Status` is now a class of it's own and it can be accessed with `cat.status`.  It can be broken up into four main properties: 
- `group` This is the group the cat is currently affiliated with. It can be None, for cats like loners, kittypets, ect.
- `rank` The rank the cat currently holds within their current group.  If they have no group, this will match their social.
- `social` How other cats regard the cat. This can be `CLAN_CAT`, `KITTYPET`, `LONER`, or `ROGUE`.
- `standing` How a group *views* the cat. A cat will have a standing with every group they have encountered (eventually this will extend to other clans, but i wanted to keep the scope of this as small as possible, so for now this is just the player clan and the afterlife groups).  This is what we use to mark a cat as exiled or lost.  Typically, cats will either be a `MEMBER` or `KNOWN`.

These all have new StrEnums to match them: `CatGroup`, `CatRank`, `CatSocial`, and `CatStanding`.  These have their own funcs as well to quickly check for standard groupings, such as if a rank is one of the apprentice ranks or if a group is an afterlife.

`CatGroup` is an interesting one here.  It has the enums: `PLAYER_CLAN`, `STAR_CLAN`, `DARK_FOREST`, `UNKNOWN_RESIDENCE`, and `OTHER_CLAN#` numbered 1-5.  The other_clan enums are a bit odd, as the number of other clans varies from save to save, but for now I have each other clan being assigned an enum at initialization.  The used enums are then saved in `game.clan.other_clans` as a list for ease of access.  Atm, I've only used it to choose a random other clan for creating new_cats that are former clancats.

Other properties of `Status` are:
- `all_socials` for a list of all socials a cat has ever been
- `all_groups` for a list of all groups the cat has been in
- `all_ranks` for a *dict* including rank as key and the moons they were that rank as the value.  This will cover all the ranks they've ever held (including those outsider ones like loner!)
- `alive_in_player_clan`, `is_outsider`, `is_clancat`, `is_former_clancat`, are all bool properties returning fairly self explanatory information

Now, all of the information for those properties is actually pulled from two lists!  These are the lists that actually get saved within the cat's save info.  
- `group_history` This is a list of dicts. Each dict contains the group, the rank, and the moons_as that rank in that group.  These dicts are in chronological order and there can be multiple dicts for matching group/rank pairings.  This is meant to allow us an explicit history of a cat's ranks.  We'll know if a cat was a warrior, then a med cat for a little bit, then a warrior again, for example.  It also means that we'll know exactly what rank a cat last held within a clan, so if they leave and then return, we'll know what they *used* be be.
- `standing_history` This is a list of dicts. Each dict contains a group, a list of standings with that group, and a bool for `near`.  Each group will have one dict within this list, no duplicates.  Standings will be in chronological order, so we'll know the entire history a cat with each group. `near` is used to dictate if the cat is within interact-able distance. Right now, this is just replacing `driven_out`.  But I'm sure you can imagine the ways in which we can expand it's use.

Of course, we have a variety of functions to help us parse all of this data in useful ways.
- `get_standing_with_group()` is an easy way to get a cat's current standing with a given group
- `find_prior_clan_rank()` lets us figure out what clan rank a cat last held within *any* clan
- `is_lost()`, `is_exiled()`, `is_near()` are funcs returning a fairly self-explanatory bool. These all take a group as an arg, so that you can check the cat's status toward a specific group.

To help us modify this this information, we have:
- `change_current_moons_as()` essentially used to retroactively change the cat's `moons_as`.  This is mostly meant for use in creating the "history" of a newly generated cat, like ensuring a newly made loner has information that makes it appear they have been a loner their whole life.
- `increase_current_moons_as()` increments `moons_as` by one, atm just used on moonskips!
- `change_standing()`, `become_lost()`, `exile_from_group()`, and `add_to_group()` are pretty self explanatory!  I expect these to see the most practical use. 
-  `_change_rank()` is a private function, generally we should always be using the greater `cat.rank_change()`, since that cat function handles all the complex moving parts of a rank change.  `._change_rank()` *only* changes the rank and the cat's resulting status information.

Everything has been provided with docstrings and I've typehinted a lot!  So this should be relatively easy to parse and use.

Did anything get yeeted? Absolutely.  
- The `Cat` class no longer has `outside`, `df`, or `exiled` attributes.  Obv, these are also no longer saved.
- `Cat.dead` and `Cat.dead_for` are now properties and their matching save file spots are gone.  Rather, we pull their info from the `Status` class!  `dead` looks to see if the cat is in an afterlife group and `dead_for` counts up the moons they've been in any afterlife group for.  They have matching `setter` methods as well, so their functionality hasn't changed at all.

As far as to what has changed within the actual gameplay, hopefully nothing! I've done my best to keep everything functioning exactly the same, though I certainly have a million ideas for things to improve now that the status class actually supports it lmao.  However, I wanted to keep the scope of this as small as possible.  The only real change users should see is that exiled and lost cats' profiles now list both "exiled/lost" *and* the social they're currently assigned ("kittypet/loner").  Save files should convert safely and mostly accurately, beyond weird edge cases or hyper old saves.
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage senior developers from merging your PR! -->

## Why This Is Good For ClanGen
There are *so* many ways we want to expand the game that were either impossible or would require such convoluted handling of the old status attribute and bool attributes that they were ill-advised.  This class fixes that issue.  We now have much more capability to manipulate cats and their position within the world.  So many things are now simpler to manage (i got rid of so many `if cat.outside and cat.exiled and cat.dead ect ect` statements) now that we aren't juggling a million bools in an attempt to keep the cat's identity straight.
<!-- If this is a bug fix, you can remove this section. -->
<!-- Please add a short description of why you think these changes would benefit the game. -->
<!-- If this PR was the result of discussion/testing on the discord, please add a link to the discord conversation here. -->
<!-- If you have multiple features that can stand on their own, or unrelated bugfixes, please create separate PRs for them. -->

## Proof of Testing
Currently have a beta tester thread running here: https://discord.com/channels/1003759225522110524/1386426068147372084 to help make sure this is thoroughly tested since I made some pretty wide-ranging changes.  Currently appears that most bugs have been taken care of.

<!-- Include any screenshots, debugging steps, or links to beta testing threads here. At least one form of proof of testing is REQUIRED for all new content. You must be able to run the code locally before you PR it here. -->

scribblecrumb committed:
adding rank enums

scribblecrumb committed:
adding social lookup json

allows us to associate certain ranks with certain social categories

scribblecrumb committed:
Social and Standing enums

scribblecrumb committed:
starting the status class

scribblecrumb committed:
update sort order to use rank enums

scribblecrumb committed:
adding status_dict TypedDict

scribblecrumb committed:
cat init new status integration

scribblecrumb committed:
handle sending cats to afterlife

scribblecrumb committed:
update exiling

scribblecrumb committed:
implement saved cat loading

scribblecrumb committed:
handle losing cats and adding cats to player clan

scribblecrumb committed:
documentation

scribblecrumb committed:
handle rank changes and some function renaming

scribblecrumb committed:
integrating status into one_moon, thoughts, rel changes

scribblecrumb committed:
Merge remote-tracking branch 'upstream/development' into status-update

scribblecrumb committed:
conditions integration

scribblecrumb committed:
more condition stuff

scribblecrumb committed:
more integration of new status into cat

scribblecrumb committed:
renaming cat enums

scribblecrumb committed:
removing un-needed cat attributes

these will be replaced by status properties

scribblecrumb committed:
_start_group_history documentation

scribblecrumb committed:
small adjustments for readability

scribblecrumb committed:
group enums

scribblecrumb committed:
add func to retrieve standing list

scribblecrumb committed:
handle exiling

scribblecrumb committed:
handle adding to clan

scribblecrumb committed:
CatGroup conversions and is_valid_mentor adjustment

scribblecrumb committed:
update_mentor conversion

scribblecrumb committed:
mediate_relationship conversion

scribblecrumb committed:
sorting and save dict conversion

scribblecrumb committed:
example cat creation

scribblecrumb committed:
some cleanup and docs

scribblecrumb committed:
convert history to new system

scribblecrumb committed:
converting names to new status

scribblecrumb committed:
converting remaining .outside, .exiled, and .df

scribblecrumb committed:
convert skills.py to new status

scribblecrumb committed:
convert thoughts.py to new status

scribblecrumb committed:
convert thoughts.py to new status

scribblecrumb committed:
converting cat_relations folder to new status

scribblecrumb committed:
convert clan_resources to new status

scribblecrumb committed:
adding is_meddie CatRank

scribblecrumb committed:
change is_any_apprentice to enum func

scribblecrumb committed:
adjust more catrank funcs

scribblecrumb committed:
debug converted to new status

scribblecrumb committed:
patrol scripts converted to new status

scribblecrumb committed:
change how status is initialized

scribblecrumb committed:
group_events.py status conversion

scribblecrumb committed:
pregnancy_events.py status conversion

scribblecrumb committed:
relation_events status conversion

scribblecrumb committed:
romantic_events status conversion

scribblecrumb committed:
welcoming_events status conversion

scribblecrumb committed:
condition_events status conversion

scribblecrumb committed:
handle_short_events status conversion

scribblecrumb committed:
update event_filter status

scribblecrumb committed:
update generate_events status

scribblecrumb committed:
update outsider_events status

scribblecrumb committed:
small status class adjustments

scribblecrumb committed:
load_cat status conversions

scribblecrumb committed:
update windows to new status

scribblecrumb committed:
update allegiance screen to new status

scribblecrumb committed:
update ceremony screen to new status

scribblecrumb committed:
update adoptive parent screen to new status

scribblecrumb committed:
update choose mate screen to new status

scribblecrumb committed:
mass replace apps and warrior CatRanks

scribblecrumb committed:
mass replace apps and warrior CatRank

scribblecrumb committed:
mass replace medicine cat CatRank

scribblecrumb committed:
mass replace mediator CatRank

scribblecrumb committed:
mass replace deputy CatRank

scribblecrumb committed:
mass replace leader CatRank

scribblecrumb committed:
mass replace elder CatRank

scribblecrumb committed:
mass replace kitten CatRank

scribblecrumb committed:
mass replace newborn CatRank

scribblecrumb committed:
replacing with is_baby where possible

scribblecrumb committed:
replacing with is_any_medicine_rank where possible

scribblecrumb committed:
replacing with is_any_warrior_like_rank where possible

scribblecrumb committed:
replacing with is_any_mediator_rank where possible

scribblecrumb committed:
replacing with is_any_mediator_rank where possible

also changed name of is_any_warrior_like_rank to is_any_adult_warrior_like_rank to make exclusion of apprentices more apparent

scribblecrumb committed:
finishing choose mentor update

scribblecrumb committed:
update clan screen to new status

scribblecrumb committed:
update clan settings screen to new status

scribblecrumb committed:
update clearing screen to new status

scribblecrumb committed:
updating leader den screen to new status

also added is_near and adjust is_lost

scribblecrumb committed:
some adjustments

fixing all_clancat_ranks use, adding some CatGroup type assignments, and making target functional on moved_to_afterlife

scribblecrumb committed:
replacing relevant lines with .in_player_clan()

kills SO many dead and outside checks

scribblecrumb committed:
update list screen to new status

scribblecrumb committed:
update med den screen to new status

scribblecrumb committed:
update mediation screen to new status

scribblecrumb committed:
update patrol screen to new status

added enum func for is_allowed_to_patrol

scribblecrumb committed:
add is_afterlife and is_clan_group

scribblecrumb committed:
documentation and small adjustments to status funcs

scribblecrumb committed:
updating profile screen to new status

scribblecrumb committed:
update role screen to new status

scribblecrumb committed:
update sprite inspect to new status

scribblecrumb committed:
update warrior den to new status

scribblecrumb committed:
update clan.py to new status

scribblecrumb committed:
update conditions.py to new status

scribblecrumb committed:
update events.py to new status

scribblecrumb committed:
updating utility to new status

scribblecrumb committed:
integrate enum assignments for other clans

scribblecrumb committed:
dead and dead_for properties

scribblecrumb committed:
Merge remote-tracking branch 'upstream/development' into status-update

scribblecrumb committed:
oopsie add update_mask back in

got yeeted in my merge

scribblecrumb committed:
remove property assignments from init

scribblecrumb committed:
fixing a bunch of issues that were crashing the game

love cleaning up my own messes

scribblecrumb committed:
fix rel loading issue

scribblecrumb committed:
fix list screen  merge

scribblecrumb committed:
fix window merge

scribblecrumb committed:
fix dead cat invasion

scribblecrumb committed:
fix UI bugs

scribblecrumb committed:
add moons_as incrementing

scribblecrumb committed:
fix clan merge

scribblecrumb committed:
fix some mistyped status stuff

scribblecrumb committed:
adjust how all_groups works

scribblecrumb committed:
fixing variety of status bugs

scribblecrumb committed:
fixing social enums and enum assignments on load

scribblecrumb committed:
adding dead checks where needed

scribblecrumb committed:
ensuring met outsider are known

scribblecrumb committed:
changes to change_standing

scribblecrumb committed:
fixing example cat creation

scribblecrumb committed:
kill that guide dead

scribblecrumb committed:
Merge remote-tracking branch 'upstream/development' into status-update

scribblecrumb committed:
fix exiled afterlife

scribblecrumb committed:
remove unused is_dead

scribblecrumb committed:
fixing a group call

scribblecrumb committed:
organizing

scribblecrumb committed:
oopsie, put this back

scribblecrumb committed:
variety of fixes

readability of dead setter,
assignments of game.clan.name as group,
missed .outside and .exiled call,
unassigned standings,
some documentation,
changing how afterlife is assigned,
dead cat loading problems

scribblecrumb committed:
switch social_lookup to enum dict

scribblecrumb committed:
documentation

scribblecrumb committed:
prevent rank entry duplication

scribblecrumb committed:
bugfixes

cats can only be given warrior rank found from their age
fix cats joining the clan without taking a clan rank
fix a standing crash

scribblecrumb committed:
fix syntax error

scribblecrumb committed:
fix loading old save issue

scribblecrumb committed:
Merge remote-tracking branch 'upstream/development' into status-update

scribblecrumb committed:
remove old arg

scribblecrumb committed:
formatting update

scribblecrumb committed:
fix a missed .status call

scribblecrumb committed:
update freshkill test to new status

scribblecrumb committed:
fix freshkill test

scribblecrumb committed:
fix cat test

scribblecrumb committed:
fix exile invite

scribblecrumb committed:
fix damn former Clancat recognizing

scribblecrumb committed:
update test conditions to new status

scribblecrumb committed:
fix lost cats joining the clan

scribblecrumb committed:
update group interaction test to new status

scribblecrumb committed:
update interactions test to new status

scribblecrumb committed:
update patrol condition test to new status

scribblecrumb committed:
fix test_cat exiled statuses

scribblecrumb committed:
update thoughts test to new status

scribblecrumb committed:
adjust test dead_exiled

scribblecrumb committed:
update utility test to new status

scribblecrumb committed:
formatting

scribblecrumb committed:
fixing formatting

scribblecrumb committed:
fix mismerge

scribblecrumb committed:
ban mediators from patrols

scribblecrumb committed:
fix formatting

scribblecrumb committed:
fix former clancats joining

scribblecrumb committed:
trying to fix tests

scribblecrumb committed:
fix null group standings

scribblecrumb committed:
kill the former clancats again

scribblecrumb committed:
fix sneaky statuses

scribblecrumb committed:
fix misassigned status

scribblecrumb committed:
fix another mistaken replace

scribblecrumb committed:
fixing freshkill. maybe?

scribblecrumb committed:
jgynns suggestion

scribblecrumb committed:
test attempt

scribblecrumb committed:
test fixes

scribblecrumb committed:
nuking outside, sc, df, and ur game.clan lists

scribblecrumb committed:
Merge remote-tracking branch 'upstream/development' into status-update

scribblecrumb committed:
Merge remote-tracking branch 'upstream/development' into status-update

scribblecrumb committed:
test fix

scribblecrumb committed:
exiled test fix

scribblecrumb committed:
fix age enum import

j-gynn commented on scripts/cat/cats.py:
```diff
@@ -480,6 +467,35 @@ def __eq__(self, other):
+    def dead(self) -> bool:
+        if self.status.group and self.status.group.is_afterlife():
+            return True
+        return False
```
```suggestion
        return self.status.group and self.status.group.is_afterlife()
```

j-gynn commented on scripts/cat/cats.py:
```diff
@@ -480,6 +467,35 @@ def __eq__(self, other):
+    def dead(self, die: bool):
+        if die:
+            if self.status.group and self.status.group.is_afterlife():
+                return
```
I feel as though some kind of warning should be raised on this - trying to assign dead to an already dead cat is probably an indicator that something has gone wrong somewhere.

j-gynn commented on scripts/cat/cats.py:
```diff
@@ -480,6 +467,35 @@ def __eq__(self, other):
+                CatGroup.DARK_FOREST,
+            ):
+                count += entry.get("moons_as")
+        return count
```
```suggestion
        return sum(entry.get("moons_as") for entry in self.status.group_history if entry.get("group") in (CatGroup.STAR_CLAN, CatGroup.UNKNOWN_RESIDENCE, CatGroup.DARK_FOREST))
```
Double-check this one in-game, but I think that's right.

j-gynn commented on scripts/cat/cats.py:
```diff
@@ -582,9 +598,11 @@ def die(self, body: bool = True):
-        if self.status == "leader":
+        darkforest = (
+            True if game.clan.instructor.status.group == CatGroup.DARK_FOREST else False
+        )
```
```suggestion
        darkforest = game.clan.instructor.status.group == CatGroup.DARK_FOREST
```

j-gynn commented on scripts/cat/cats.py:
```diff
@@ -600,7 +618,7 @@ def die(self, body: bool = True):
                 final_thought = event_text_adjust(self, death_thought, main_cat=self)
                 self.thought = final_thought
-                if game.clan.instructor.df is False:
+                if darkforest is False:
```
```suggestion
                if not darkforest:
```

j-gynn commented on scripts/cat/cats.py:
```diff
@@ -889,26 +903,29 @@ def add_to_clan(self) -> list:
         return ids
 
-    def status_change(self, new_status, resort=False):
+    def rank_change(self, new_rank, resort=False):
```
RT `new_rank`, is this a string? I'd suggest it should be the CatRank enum...? Also the description is outdated

j-gynn commented on scripts/cat/cats.py:
```diff
@@ -970,7 +987,11 @@ def status_change(self, new_status, resort=False):
+            CatRank.WARRIOR,
+            CatRank.MEDICINE_CAT,
+            CatRank.MEDIATOR,
+        ]:
```
```suggestion
        if self.status.rank in (
            CatRank.WARRIOR,
            CatRank.MEDICINE_CAT,
            CatRank.MEDIATOR,
        ):
```

j-gynn commented on scripts/cat/enums.py:
```diff
@@ -13,11 +13,115 @@ class CatAgeEnum(StrEnum):
+    OTHER_CLAN5 = "other_clan5"
+
+    DARK_FOREST = "dark_forest"
+    STAR_CLAN = "star_clan"
```
Minor quibble, but canonically this is written "StarClan", so I'd vouch for `STARCLAN`.

scribblecrumb replied:
I'd wanted it to match our other clan enums, but tbf i was also constantly having to stop myself from just writing STARCLAN lmao

j-gynn commented on scripts/cat/cats.py:
```diff
@@ -917,18 +934,18 @@ def status_change(self, new_status, resort=False):
 
         # updates mentors
-        if self.status == "apprentice":
+        if self.status.rank == CatRank.APPRENTICE:
```
I feel like all the `pass` ranks should be grouped together...? Minor quibble.

j-gynn commented on scripts/cat/cats.py:
```diff
@@ -997,10 +1018,19 @@ def rank_change_traits_skill(self, mentor):
             History.add_mentor_facet_influence_strings(self)
         return
 
+    def change_name(self, new_prefix=None, new_suffix=None):
```
Why does this need to exist?

scribblecrumb replied:
We had this bit of code being repeated a lot in the utility create_new_cat function. Using this makes the utility function way more readable.

j-gynn commented on scripts/cat/cats.py:
```diff
@@ -1223,26 +1253,39 @@ def generate_lead_ceremony(self):
+            if self.fetch_cat(i)
+            and i not in life_givers
+            and self.fetch_cat(i).status.group == CatGroup.DARK_FOREST
+        ]
```
Why do we calculate both of these when our instructor can only be in one of them?

j-gynn commented on scripts/cat/cats.py:
```diff
@@ -3521,8 +3523,7 @@ def determine_next_and_previous_cats(
             if check_cat.dead == self.dead
-            and check_cat.outside == self.outside
-            and check_cat.df == self.df
+            and check_cat.status.group == self.status.group
```
I don't think this works with COTC, given the new OTHERCLAN groups...?

j-gynn commented on scripts/cat/names.py:
```diff
@@ -239,7 +240,7 @@ def give_suffix(self, pelt, biome, tortiepattern):
             if named_after_pelt:
                 if (
-                    pelt in ("Tortie", "Calico")
+                    pelt in ["Tortie", "Calico"]
```
Why this change...?

j-gynn commented on scripts/cat/skills.py:
```diff
@@ -106,7 +108,7 @@ def get_random(exclude: list = ()):
         uncommon_paths = [
             i
-            for i in (
+            for i in [
```
Why this change?

j-gynn commented on scripts/cat/skills.py:
```diff
@@ -495,15 +500,24 @@ def progress_skill(self, the_cat):
+                        True
+                        if the_cat.status.rank in [CatRank.APPRENTICE, CatRank.KITTEN]
+                        else False
+                    ),
```
```suggestion
                    interest_only=the_cat.status.rank in (CatRank.APPRENTICE, CatRank.KITTEN),
```

j-gynn commented on scripts/cat/skills.py:
```diff
@@ -495,15 +500,24 @@ def progress_skill(self, the_cat):
+                        True
+                        if the_cat.status.rank in [CatRank.APPRENTICE, CatRank.KITTEN]
+                        else False
+                    ),
```
```suggestion
                    interest_only=(the_cat.status.rank in (CatRank.APPRENTICE, CatRank.KITTEN)
                    ),
```

j-gynn commented on scripts/cat/skills.py:
```diff
@@ -587,7 +601,7 @@ def progress_skill(self, the_cat):
             # For outside cats, just check interest and flip it if needed.
             # Going on age, rather than status here.
-            if the_cat.age not in ("kitten", "adolescent"):
+            if the_cat.age not in ["kitten", "adolescent"]:
```
```suggestion
            if the_cat.age not in (CatAge.KITTEN, CatAge.ADOLESCENT):
```
I think...?

j-gynn commented on scripts/cat/status.py:
```diff
@@ -0,0 +1,639 @@
+        """
+
+        self.group_history = group_history if group_history else []
+        """List of dicts containing the keys group, rank, and moons_as. A new dict is added anytime group or rank are
```
```suggestion
        """List of dicts containing the keys group, rank, and moons_as. A new dict is added whenever group or rank are
```

j-gynn commented on scripts/cat/status.py:
```diff
@@ -0,0 +1,639 @@
+            for enum in CatRank:
+                if enum == entry["rank"]:
+                    entry["rank"] = enum
+                    break
```
What is this doing? If it's just parsing the save info as an Enum, there's a way of doing that baked into Enums iirc...

j-gynn commented on tests/test_cat.py:
```diff
@@ -576,14 +650,44 @@ def test_specsuffix_lost(self):
+        )
+
+        for rank in clancat_ranks:
+            with self.subTest("clancat social assignment"):
```
```suggestion
            with self.subTest("clancat social assignment", rank=rank):
```
putting this here tells us which rank in the subtest failed

j-gynn commented on scripts/utility.py:
```diff
@@ -538,14 +539,14 @@ def create_new_cat_block(
 
     # KITTEN THOUGHT
-    if status in ("kitten", "newborn"):
+    if rank in [CatRank.KITTEN, CatRank.NEWBORN]:
```
```suggestion
    if rank in (CatRank.KITTEN, CatRank.NEWBORN):
```

j-gynn commented on scripts/utility.py:
```diff
@@ -617,15 +629,14 @@ def create_new_cat_block(
-            kittypet=cat_type == "kittypet",
-            other_clan=cat_type == "former Clancat",
-            kit=(not litter) and status in ("kitten", "newborn"),
+            kit=False if litter else rank in [CatRank.KITTEN, CatRank.NEWBORN],
```
```suggestion
            kit=False if litter else rank in (CatRank.KITTEN, CatRank.NEWBORN),
```

j-gynn commented on scripts/utility.py:
```diff
@@ -785,34 +797,45 @@ def create_new_cat(
+        elif moons >= 12:
+            rank = CatRank.WARRIOR
+        elif moons >= 120:
+            rank = CatRank.ELDER
```
It's impossible to set a CatRank.ELDER here (they'll always pass the CatRank.WARRIOR check above.

j-gynn commented on scripts/utility.py:
```diff
@@ -822,88 +845,82 @@ def create_new_cat(
+            if new_cat.status.rank != rank:
+                new_cat.status.change_rank(rank)
+            # give apprentice aged cat a mentor
+            if new_cat.status.rank == CatRank.APPRENTICE:
```
Excludes medcat app and mediator app - intended?

j-gynn commented on scripts/utility.py:
```diff
@@ -822,88 +845,82 @@ def create_new_cat(
-                        status=status,
-                        gender=_gender,
-                        backstory=backstory,
-                        parent1=parent1,
-                        parent2=parent2,
-                        adoptive_parents=adoptive_parents if adoptive_parents else [],
-                    )
-            # these cats keep their old names
+                        new_cat.change_name(new_prefix=new_prefix)
+                # else, take a whole new name
+                else:
+                    new_cat.change_name()
```
as discussed before, this means that only cats with two word or more loner names can keep part of it. Intended?

scribblecrumb replied:
This one doesn't do that, you're thinking of the create_new_cat_block function, which *is* broken lmao, I'll go ahead and fix that one

j-gynn commented on scripts/utility.py:
```diff
@@ -962,13 +976,12 @@ def create_new_cat(
 
                 # assign scars
-                if chosen_condition in ("lost a leg", "born without a leg"):
+                if chosen_condition in ["lost a leg", "born without a leg"]:
```
Why?

j-gynn commented on scripts/utility.py:
```diff
@@ -962,13 +976,12 @@ def create_new_cat(
+                if chosen_condition in ["lost a leg", "born without a leg"]:
                     new_cat.pelt.scars.append("NOPAW")
-                elif chosen_condition in ("lost their tail", "born without a tail"):
+                elif chosen_condition in ["lost their tail", "born without a tail"]:
```
Why?

j-gynn commented on scripts/screens/Screens.py:
```diff
@@ -900,12 +901,10 @@ def set_cat_location_bg(self, cat, bg: str = "default"):
+                if cat.status.group == CatGroup.DARK_FOREST
+                else "unknown_residence"
+                if cat.status.group == CatGroup.UNKNOWN_RESIDENCE
+                else "starclan"
```
We could probably make this more optimal now given that we can explicitly check for SC (most common instance, previously had to be determined by process of elimination).

j-gynn commented on scripts/screens/Screens.py:
```diff
@@ -68,10 +69,10 @@ def change_screen(self, new_screen):
             game.last_list_forProfile = None
-            self.current_group = "your_clan"
-            self.death_status = "living"
+            self.current_group = "clan"
```
What does this change...?

j-gynn commented:
I'm working on reviewing this, but whilst I do - there's a few conflicts that need resolving!

scribblecrumb committed:
Merge remote-tracking branch 'upstream/development' into status-update

j-gynn commented on scripts/cat/thoughts.py:
```diff
@@ -326,31 +322,22 @@ def create_thoughts(
-        elif main_cat.outside:
+        elif main_cat.status.is_outsider():
             spec_dir = "/alive_outside"
         else:
```
Shouldn't this be based on CatGroup now?

j-gynn commented on scripts/conditions.py:
```diff
@@ -41,7 +43,7 @@ def amount_clanmembers_covered(all_cats, amount_per_med) -> int:
         elif cat.skills.meets_skill_requirement(SkillPath.HEALER, 2):
             total_med_number += 1.75
-        elif cat.skills.meets_skill_requirement(SkillPath.HEALER, 1):
+        elif cat.skills.meets_skill_requirement(SkillPath.HEALER, 2):
```
```suggestion
        elif cat.skills.meets_skill_requirement(SkillPath.HEALER, 1):
```
Merge mistake

j-gynn commented on scripts/cat/status.py:
```diff
@@ -0,0 +1,639 @@
+            if record["group"] not in groups:
+                groups.append(record["group"])
+
+        return [record["group"] for record in self.group_history]
```
Uhh, looks like we're not correctly using `lines 272-276` here?

j-gynn commented on scripts/cat/status.py:
```diff
@@ -0,0 +1,639 @@
+        rank = [
+            rank for rank in list(CatRank) if rank == self.group_history[-1]["rank"]
+        ]
+        return rank[0]
```
```suggestion
        return CatRank(self.group_history[-1]["rank"])
```
I believe should work? I can't find documentation on what error this throws if it's not a valid Enum, but in fairness it probably should crash if it's invalid in this case anyway

j-gynn commented on scripts/cat/status.py:
```diff
@@ -0,0 +1,639 @@
+                ]
+            )
+        elif age in [CatAge.YOUNG_ADULT, CatAge.ADULT, CatAge.SENIOR_ADULT]:
+            rank = CatRank.WARRIOR
```
No chance to be a mediator or medicine cat?

edit: I see this is used for deputy & leader being reinvited to the Clan when they're exiled, perhaps that's why. Still feels strange that you can be randomly assigned any of the apprentice paths but only warrior path if you're an adult.

scribblecrumb replied:
Fair point, I've made `add_to_group` just assign the returning lead/dep to warrior or elder depending on their age. So this function can now give all of those adult ranks.

j-gynn commented on scripts/cat/status.py:
```diff
@@ -0,0 +1,639 @@
+            rank = choice(
+                [
+                    CatRank.APPRENTICE,
+                    CatRank.MEDIATOR_APPRENTICE,
```
What if the player has mediators disabled?

scribblecrumb replied:
I don't think players can actually disable mediators completely.  The setting is just to disallow warriors/elders from becoming mediators over a timeskip, it doesn't disallow mediators from generating.  idk what it is we're intending that setting to accomplish, so this might be smth the team needs to discuss.

j-gynn commented on scripts/cat/status.py:
```diff
@@ -0,0 +1,639 @@
+
+    def change_standing(self, new_standing: CatStanding, group: CatGroup = None):
+        """
+        Update the given group with the given standing. If not group is given, the new standing will be added to the
```
```suggestion
        Update the given group with the given standing. If no group is given, the new standing will be added to the
```

j-gynn commented on scripts/cat/status.py:
```diff
@@ -0,0 +1,639 @@
+
+        if not group:
+            group = self.group
+        found_record = False
```
```suggestion
```
This and the following 2 suggestions remove the flag `found_record` and just make use of returning early

j-gynn commented on scripts/cat/status.py:
```diff
@@ -0,0 +1,639 @@
+        for record in self.standing_history:
+            if record["group"] == group:
+                record["standing"].append(new_standing)
+                found_record = True
```
```suggestion
                return
```

j-gynn commented on scripts/cat/status.py:
```diff
@@ -0,0 +1,639 @@
+        if not found_record:
+            self.standing_history.append(
+                {"group": group, "standing": [new_standing], "near": True}
+            )
```
```suggestion
        self.standing_history.append(
            {"group": group, "standing": [new_standing], "near": True}
        )
```

j-gynn commented on scripts/cat/status.py:
```diff
@@ -0,0 +1,639 @@
+        for enum in CatRank:
+            if enum == new_social_status:
+                rank = enum
+                break
```
I think you can do this programmatically without having to loop through all the Enums...

j-gynn commented on scripts/cat/status.py:
```diff
@@ -0,0 +1,639 @@
+
+    def change_rank(self, new_rank: CatRank):
+        """
+        Changes the cats rank to the new_rank. Generally you shouldn't use just this to change a cat's rank!
```
If this shouldn't be used, could it be made private with an underscore? Technically that's not actually private since python doesn't have the concept, but it should highlight in IDEs that you're doing a Thing Not Allowed

It's quite easy to miss the comment, harder to miss a yellow wiggly line ;)

j-gynn commented on scripts/cat/status.py:
```diff
@@ -0,0 +1,639 @@
+        Flips the "near" bool of the given group.
+        """
+        for entry in self.standing_history:
+            if entry.get("group") == group:
```
Should it `return` after finding this?

j-gynn commented on scripts/cat/status.py:
```diff
@@ -0,0 +1,639 @@
+        """
+        for entry in self.standing_history:
+            if entry.get("group") == group:
+                if entry.get("near"):
+                    entry["near"] = False
+                else:
+                    entry["near"] = True
```
```suggestion
                if entry.get("near"):
                    entry["near"] = not entry.get("near", False)
```
Also, what cases are there groups that don't have a `near` value? Is that Playerclan?

scribblecrumb replied:
There's actually no case where they can't have a near, so I guess it makes more sense to let it crash if that's missing somehow, there would def be something wrong going on.

j-gynn commented on scripts/cat/status.py:
```diff
@@ -0,0 +1,639 @@
+                standing_list = entry["standing"]
+                break
+
+        return standing_list
```
```suggestion
        return [entry["standing"] for entry in self.standing_history if entry["group"] == group]
```

j-gynn commented on scripts/cat/status.py:
```diff
@@ -0,0 +1,639 @@
+            if CatSocial.CLANCAT in self.all_socials
+            and self.social != CatSocial.CLANCAT
+            else False
+        )
```
```suggestion
        return (CatSocial.CLANCAT in self.all_socials
            and self.social != CatSocial.CLANCAT)
```
May not need the parentheses, might accidentally make that into a tuple 😬 please check

j-gynn commented on scripts/cat/status.py:
```diff
@@ -0,0 +1,639 @@
+        if standing and standing[-1] == CatStanding.EXILED:
+            return True
+
+        return False
```
```suggestion
        return standing and standing[-1] == CatStanding.EXILED:
```

j-gynn commented on scripts/events_module/outsider_events.py:
```diff
@@ -44,32 +38,15 @@ def killing_outsiders(cat: Cat):
                     )
+                else:
+                    text = (
+                        f"Rumors reach your Clan that the {cat.status.social} "
```
Just as an aside, does this make former Clancat format correctly? Former Clancat my beloathed

j-gynn commented on scripts/events_module/patrol/patrol.py:
```diff
@@ -173,32 +161,32 @@ def add_patrol_cats(self, patrol_cats: List[Cat], clan: Clan) -> None:
 
-            # Combined patrol_statuses categories
-            if cat.status in ("medicine cat", "medicine cat apprentice"):
+            # Combined patrol_statuses catagories
```
```suggestion
            # Combined patrol_statuses categories
```
Merge mistake

j-gynn commented on scripts/events_module/patrol/patrol.py:
```diff
@@ -454,11 +437,14 @@ def get_possible_patrols(
+                if (
+                    _pat.patrol_id
+                    == game.config["patrol_generation"]["debug_ensure_patrol_id"]
+                ):
```
Congrats! You've merged in an entire PR incorrectly! \*sad kazoo*

Please go back and reintroduce the fix <3

scribblecrumb replied:
i. i think i've fixed it ;-;

j-gynn commented on scripts/events_module/patrol/patrol.py:
```diff
@@ -483,10 +469,6 @@ def _check_constraints(self, patrol: PatrolEvent) -> bool:
-                print(
-                    "DEBUG: requested patrol does not meet constraints (relationship type)"
-                )
             return False
```
Revert change.

j-gynn commented on scripts/events_module/patrol/patrol.py:
```diff
@@ -472,7 +458,7 @@ def get_possible_patrols(
                     f"debug_ensure_patrol_id: "
                     f'"{game.config["patrol_generation"]["debug_ensure_patrol_id"]}" '
-                    f"is not found. Check output for reason."
+                    f"is not found."
```
Revert change.

j-gynn commented on scripts/events_module/patrol/patrol.py:
```diff
@@ -436,12 +425,6 @@ def get_possible_patrols(
-            print(
-                "DEBUG: requested patrol not present (check spelling/mismatched season, biome, patrol type, new cat flag, other clan relations, disaster setting)"
-            )
-
```
Revert change.

j-gynn commented on scripts/events_module/patrol/patrol.py:
```diff
@@ -495,16 +477,12 @@ def _check_constraints(self, patrol: PatrolEvent) -> bool:
             )
         ):
-            if self.debug_patrol and self.debug_patrol == patrol.patrol_id:
-                print("DEBUG: requested patrol does not meet constraints (pl_skill)")
```
Revert change.

j-gynn commented on scripts/events_module/patrol/patrol.py:
```diff
@@ -495,16 +477,12 @@ def _check_constraints(self, patrol: PatrolEvent) -> bool:
             and self.patrol_leader.personality.trait not in patrol.pl_trait_constraints
         ):
-            if self.debug_patrol and self.debug_patrol == patrol.patrol_id:
-                print("DEBUG: requested patrol does not meet constraints (pl_trait)")
```
Revert change.

j-gynn commented on scripts/events_module/patrol/patrol.py:
```diff
@@ -561,7 +539,7 @@ def decide_if_romantic(
             value_check = check_relationship_value(love1, love2, val)
             if (
-                val in ("romantic", "platonic", "admiration", "comfortable", "trust")
+                val in ["romantic", "platonic", "admiration", "comfortable", "trust"]
```
Revert change.

j-gynn commented on scripts/events_module/patrol/patrol.py:
```diff
@@ -602,10 +580,6 @@ def _filter_patrols(
-                if self.debug_patrol and self.debug_patrol == patrol.patrol_id:
-                    print(
-                        "DEBUG: requested patrol does not meet constraints (min or max cats range)"
-                    )
```
Revert change.

j-gynn commented on scripts/events_module/patrol/patrol.py:
```diff
@@ -618,53 +592,25 @@ def _filter_patrols(
-                if self.debug_patrol and self.debug_patrol == patrol.patrol_id:
-                    print(
-                        "DEBUG: requested patrol does not meet constraints (min max status)"
-                    )
```
Revert change.

j-gynn commented on scripts/events_module/patrol/patrol.py:
```diff
@@ -618,53 +592,25 @@ def _filter_patrols(
 
             if not event_for_tags(patrol.tags, Cat):
-                if self.debug_patrol and self.debug_patrol == patrol.patrol_id:
-                    print("DEBUG: requested patrol does not meet constraints (tags)")
```
Revert change.

j-gynn commented on scripts/events_module/patrol/patrol.py:
```diff
@@ -618,53 +592,25 @@ def _filter_patrols(
 
             if biome not in patrol.biome and "any" not in patrol.biome:
-                if self.debug_patrol and self.debug_patrol == patrol.patrol_id:
-                    print("DEBUG: requested patrol does not meet constraints (biome)")
```
Revert change.

j-gynn commented on scripts/events_module/patrol/patrol.py:
```diff
@@ -618,53 +592,25 @@ def _filter_patrols(
                 continue
             if camp not in patrol.camp and "any" not in patrol.camp:
-                if self.debug_patrol and self.debug_patrol == patrol.patrol_id:
-                    print("DEBUG: requested patrol does not meet constraints (camp)")
```
Revert change.

j-gynn commented on scripts/events_module/patrol/patrol.py:
```diff
@@ -618,53 +592,25 @@ def _filter_patrols(
                 continue
             if current_season not in patrol.season and "any" not in patrol.season:
-                if self.debug_patrol and self.debug_patrol == patrol.patrol_id:
-                    print("DEBUG: requested patrol does not meet constraints (season)")
```
Revert change.

j-gynn commented on scripts/events_module/patrol/patrol.py:
```diff
@@ -618,53 +592,25 @@ def _filter_patrols(
-                if self.debug_patrol and self.debug_patrol == patrol.patrol_id:
-                    print(
-                        "DEBUG: requested patrol does not meet constraints (patrol type)"
-                    )
```
Revert change.

j-gynn commented on scripts/events_module/patrol/patrol.py:
```diff
@@ -618,53 +592,25 @@ def _filter_patrols(
-                if self.debug_patrol and self.debug_patrol == patrol.patrol_id:
-                    print(
-                        "DEBUG: requested patrol does not meet constraints (patrol type)"
-                    )
```
Revert change.

j-gynn commented on scripts/events_module/patrol/patrol.py:
```diff
@@ -618,53 +592,25 @@ def _filter_patrols(
-                if self.debug_patrol and self.debug_patrol == patrol.patrol_id:
-                    print(
-                        "DEBUG: requested patrol does not meet constraints (patrol type)"
-                    )
```
Revert change.

j-gynn commented on scripts/events_module/patrol/patrol.py:
```diff
@@ -618,53 +592,25 @@ def _filter_patrols(
-                if self.debug_patrol and self.debug_patrol == patrol.patrol_id:
-                    print(
-                        "DEBUG: requested patrol does not meet constraints (patrol type)"
-                    )
```
Revert change.

j-gynn commented on scripts/events_module/patrol/patrol.py:
```diff
@@ -710,13 +656,6 @@ def get_filtered_patrols(
-                    print(
-                        "DEBUG: requested patrol removed during herb filtering (not target herb)"
-                    )
-
```
Revert change.

j-gynn commented on scripts/events_module/patrol/patrol.py:
```diff
@@ -978,10 +917,6 @@ def balance_hunting(self, possible_patrols: list):
-            elif self.debug_patrol and self.debug_patrol == patrol.patrol_id:
-                print(
-                    "DEBUG: requested patrol does not meet constraints (failed prey balancing)"
-                )
```
Revert change.

j-gynn commented on scripts/events_module/patrol/patrol.py:
```diff
@@ -90,12 +84,6 @@ def setup_patrol(self, patrol_cats: List[Cat], patrol_type: str) -> str:
-            if game.config["patrol_generation"]["debug_ensure_patrol_id"]
-            else False
-        )
-
```
Revert change.

j-gynn commented on scripts/events_module/patrol/patrol.py:
```diff
@@ -59,9 +56,6 @@ def __init__(self):
 
-        # False if no debug patrol set, value if one is set
-        self.debug_patrol: Union[bool, str] = False
-
```
Revert change.

j-gynn commented on scripts/events_module/patrol/patrol.py:
```diff
@@ -30,8 +29,6 @@
     adjust_list_text,
 )
 
-logger = logging.getLogger(__name__)
```
Revert change.

j-gynn commented on scripts/events_module/patrol/patrol.py:
```diff
@@ -1,17 +1,16 @@
 #!/usr/bin/env python3
 # -*- coding: ascii -*-
-import logging
```
Revert change.

j-gynn commented on scripts/events_module/patrol/patrol_outcome.py:
```diff
@@ -460,7 +461,7 @@ def _handle_exp(self, patrol: "Patrol") -> str:
 
         base_exp = 0
-        if "master" in (x.experience_level for x in patrol.patrol_cats):
+        if "master" in [x.experience_level for x in patrol.patrol_cats]:
```
Revert change.

j-gynn commented on scripts/events_module/relationship/pregnancy_events.py:
```diff
@@ -147,7 +141,7 @@ def handle_having_kits(cat, clan):
         """Handle if the there is no pregnancy but the pair triggered kits chance."""
         if other_cat and (
-            other_cat.dead or other_cat.outside or other_cat.birth_cooldown > 0
+            not other_cat.status.in_player_clan() or other_cat.birth_cooldown > 0
```
I think this is more specifically trying to check whether the cat is present (this doesn't account for lost/exiled clancats I don't think?)

scribblecrumb replied:
This would still be accomplishing that tho.  `in_player_clan()` would return false if the cat was lost or exiled.

j-gynn replied:
I think I was having a fundamental misunderstanding of the scope of what `in_player_clan` means at this point lol. My mind was still considering lost and exiled cats to be part of the Clan.

j-gynn commented on scripts/events_module/relationship/pregnancy_events.py:
```diff
@@ -213,7 +207,7 @@ def handle_adoption(cat: Cat, other_cat=None, clan=game.clan):
         """Handles if the cat is zero moons pregnant."""
         if other_cat and (
-            other_cat.dead or other_cat.outside or other_cat.birth_cooldown > 0
+            not other_cat.status.in_player_clan() or other_cat.birth_cooldown > 0
```
Same as previous (not accounting for lost/exiled)

j-gynn commented on scripts/events_module/relationship/pregnancy_events.py:
```diff
@@ -425,20 +424,25 @@ def handle_two_moon_pregnant(cat: Cat, clan=game.clan):
+        ):
             involved_cats.append(other_cat.ID)
             cat_dict["r_c"] = other_cat
+            # TODO: this seems odd, outsider mates are also treated as dead?
```
Hey, can u make an issue for this pls, I agree it's strange

j-gynn commented on scripts/events_module/relationship/pregnancy_events.py:
```diff
@@ -809,36 +815,28 @@ def get_kits(
                     blood_parent = create_new_cat(
                         Cat,
-                        status=random.choice(["loner", "kittypet"]),
+                        rank=random.choice(["loner", "kittypet"]),
```
does this parse correctly? I'm sure I'll come back after I've read utility.py and be like "heh idiot ofc it's fine"

j-gynn commented on scripts/events_module/relationship/pregnancy_events.py:
```diff
@@ -1108,11 +1102,7 @@ def get_balanced_kit_chance(
-                for i in Cat.all_cats.values()
-                if not i.dead and not i.outside and not i.exiled
-            ]
+            [i for i in Cat.all_cats.values() if i.status.in_player_clan()]
```
does `in_player_clan` handle exiled and lost and w/e?

j-gynn commented on scripts/events_module/relationship/pregnancy_events.py:
```diff
@@ -1178,7 +1168,7 @@ def get_balanced_kit_chance(
         # AGE
         # - decrease the inverse chance if the whole clan is really old
-        avg_age = int(sum(cat.moons for cat in Cat.all_cats.values()) / living_cats)
+        avg_age = int(sum([cat.moons for cat in Cat.all_cats.values()]) / living_cats)
```
Revert change.

j-gynn commented on scripts/events_module/relationship/romantic_events.py:
```diff
@@ -668,7 +669,7 @@ def check_if_breakup(cat_from, cat_to):
 
         # Moving on, not breakups, occur when one mate is dead or outside.
-        if cat_from.dead or cat_from.outside or cat_to.dead or cat_to.outside:
+        if not cat_from.status.in_player_clan() or not cat_to.status.in_player_clan():
```
This confuses me since we're no longer explicitly checking for death. I know it's implicit in the system that a cat is only `in_player_clan` in life, but technically their affiliation to their Clan shouldn't end in death? They are still PlayerClan cats, they're just... dead. From a readability perspective, `in_player_clan` feels like it should return `True` even if the cat is dead, imo. Maybe it needs renaming...? `alive_in_player_clan` maybe? or something like that. Because otherwise it's easy to forget the implicit "alive" check that also happens here.

scribblecrumb replied:
mm, fair point, I'll go ahead and rename it

j-gynn commented on scripts/events_module/short/handle_short_events.py:
```diff
@@ -229,7 +223,7 @@ def handle_event(
         # this kind of sucks tho it would be nice to change how this sort of thing is handled
         if "kit_manipulated" in self.chosen_event.tags:
-            kit = Cat.fetch_cat(choice(get_alive_status_cats(Cat, ["kitten"])))
+            kit = Cat.fetch_cat(choice(find_alive_cats_with_rank(Cat, ["kitten"])))
```
should use CatRank enum.

j-gynn commented on scripts/events_module/short/handle_short_events.py:
```diff
@@ -190,21 +191,14 @@ def handle_event(
-                if n_c[0].ID in self.involved_cats:
-                    self.involved_cats.remove(str(n_c[0].ID))
+        # create new cats (must happen here so that new cats can be included in further changes)
+        self.handle_new_cats()
```
Why was this moved? /genq

j-gynn commented on scripts/events_module/short/handle_short_events.py:
```diff
@@ -391,12 +385,8 @@ def handle_new_cats(self):
-                        or "unknown" in attribute_list
-                    ):
+                elif cat.status.is_outsider():
+                    if "unknown" in attribute_list:
```
Can't exclude new cats from displaying in the event any more...?

j-gynn commented on scripts/events_module/short/handle_short_events.py:
```diff
@@ -444,7 +434,7 @@ def handle_accessories(self, pelts=Pelt):
 
         for acc in possible_accs:
-            if acc not in ("WILD", "PLANT", "COLLAR"):
+            if acc not in ["WILD", "PLANT", "COLLAR"]:
```
Revert change.

j-gynn commented on scripts/game_structure/load_cat.py:
```diff
@@ -137,7 +154,7 @@ def json_load():
             )
 
-            # Runs a bunch of apperence-related convertion of old stuff.
+            # Runs a bunch of appearance-related convertion of old stuff.
```
```suggestion
            # Runs a bunch of appearance-related conversion of old stuff.
```
since we're here...

j-gynn commented on scripts/game_structure/load_cat.py:
```diff
@@ -215,14 +231,36 @@ def json_load():
-            new_cat.dead = cat["dead"]
+
+            # checking for old dead
+            if cat.get("dead"):
```
```suggestion
            if cat.get("dead") or cat.get("df"): 
```
Combine the block below with this one so they're both handled at the same time

j-gynn commented on scripts/game_structure/load_cat.py:
```diff
@@ -215,14 +231,36 @@ def json_load():
-            new_cat.outside = cat["outside"] if "outside" in cat else False
+            # these should properly change the cat's status to align with old bool info
+            if cat.get("df"):
+                new_cat.status.send_to_afterlife(target=CatGroup.DARK_FOREST)
```
Remove as it should be covered by the above block

j-gynn commented on scripts/game_structure/load_cat.py:
```diff
@@ -490,7 +529,7 @@ def csv_load(all_cats):
                 app = Cat.all_cats.get(app_id)
                 # Make sure if cat isn't an apprentice, they're a former apprentice
-                if "apprentice" in app.status:
+                if CatRank.APPRENTICE == app.status.rank:
```
```suggestion
                if app.status.rank == CatRank.APPRENTICE:
``` 
Minor thing but the changing value should always be first

j-gynn commented on scripts/cat/status.py:
With how often I've seen `if cat.status.rank == CatRank.LEADER`, I think it might be good to add a property here. `if cat.status.is_leader` just feels cleaner imo. Feel free to disregard, but it'd also mean we don't have to keep importing the CatRank enum just for this one check lol

j-gynn commented on scripts/screens/ClanSettingsScreen.py:
```diff
@@ -395,34 +396,34 @@ def open_clan_stats(self):
+                elif cat.status.group == CatGroup.UNKNOWN_RESIDENCE:
                     ur += 1
                 else:
                     starclan += 1
```
minor optimisation thing here - since most players keep their cats going to starclan, the order of most frequent tends to be starclan -> unknown residence -> dark forest. It'd be nice if this order could be reflected in the checks. But this is *super* minor, and I don't think anyone's ever complained of lag opening the clan stats screen, but you never know.

j-gynn commented on scripts/screens/ClearingScreen.py:
```diff
@@ -93,7 +94,7 @@ def handle_event(self, event):
                 if self.focus_name:
                     self.focus_name.kill()
-            elif event.ui_element in (self.feed_one_button, self.feed_max_button):
+            elif event.ui_element in [self.feed_one_button, self.feed_max_button]:
```
Revert change.

j-gynn commented on scripts/screens/ClearingScreen.py:
```diff
@@ -795,8 +796,8 @@ def create_checkboxes(self):
-                        f"general.{status}",
-                        count=2 if status not in ("leader", "deputy") else 1,
+                        f"general.{rank}",
+                        count=2 if rank not in [CatRank.LEADER, CatRank.DEPUTY] else 1,
```
```suggestion
                        count=2 if rank not in (CatRank.LEADER, CatRank.DEPUTY) else 1,
```

j-gynn commented on scripts/screens/LeaderDenScreen.py:
```diff
@@ -149,7 +150,11 @@ def screen_switches(self):
         self.no_leader = False
 
-        if not game.clan.leader or game.clan.leader.dead or game.clan.leader.exiled:
+        if (
+            not game.clan.leader
+            or game.clan.leader.dead
+            or game.clan.leader.status.is_outsider()
+        ):
```
this is making me 🤨 but I can't place why

scribblecrumb replied:
changed it to just a .alive_in_player_clan check since that covers both dead and outsider

j-gynn commented on scripts/screens/LeaderDenScreen.py:
```diff
@@ -988,7 +987,9 @@ def update_outsider_cats(self):
             i
             for i in Cat.all_cats.values()
-            if i.outside and not i.dead and not i.driven_out
+            if not i.dead
+            and i.status.is_outsider()
```
Inconsistent behaviour relative to `cat.status.in_player_clan`, the assumed parallel function.

Am I to take it that `is_outsider` does **not** check whether the cat is dead, but `in_player_clan` does?

scribblecrumb replied:
I guess the important distinction here is that `PLAYER_CLAN` is a `CatGroup`, while outsider is just any `CatStanding` that isn't `CLANCAT`.  Since the afterlives are their own `CatGroups`, once a cat has died they're no longer part of any living `CatGroup`, including their clan.  However, a cat could be dead without being a `CLANCAT`, thus they're still outsiders.

j-gynn commented on scripts/screens/ListScreen.py:
```diff
@@ -726,13 +735,7 @@ def get_sc_cats(self):
-                and not the_cat.df
-                and not the_cat.faded
-            ):
+            if the_cat.status.group == CatGroup.STAR_CLAN and not the_cat.faded:
```
No longer checking for instructor ID here...?

j-gynn commented on scripts/screens/ListScreen.py:
```diff
@@ -744,12 +747,7 @@ def get_df_cats(self):
-                and the_cat.df
-                and not the_cat.faded
-            ):
+            if the_cat.status.group == CatGroup.DARK_FOREST and not the_cat.faded:
```
No longer checking for instructor ID here...?

j-gynn commented on scripts/screens/ListScreen.py:
```diff
@@ -760,9 +758,5 @@ def get_ur_cats(self):
-                and not the_cat.faded
-                and not the_cat.driven_out
-            ):
+            if the_cat.status.group == CatGroup.UNKNOWN_RESIDENCE and not the_cat.faded:
```
Missing the `is_near` check.
Assuming that, upon death, cats you drove away not be reappearing in the list...?

j-gynn commented on scripts/screens/MediationScreen.py:
```diff
@@ -111,10 +111,10 @@ def handle_event(self, event):
                     self.selected_cat_1,
                     self.selected_cat_2,
-                ):
+                ]:
```
Revert change.

j-gynn commented on scripts/screens/MediationScreen.py:
```diff
@@ -111,10 +111,10 @@ def handle_event(self, event):
                 self.update_selected_cats()
             elif event.ui_element in self.cat_buttons:
-                if event.ui_element.return_cat_object() not in (
+                if event.ui_element.return_cat_object() not in [
```
Revert change.

j-gynn commented on scripts/screens/MediationScreen.py:
```diff
@@ -567,16 +565,16 @@ def draw_info_block(self, cat, starting_pos: tuple):
+                if cat.genderalign in ["female", "trans female"]:
                     col2 += i18n.t("general.aunt")
-                elif cat.genderalign in ("male", "trans male"):
+                elif cat.genderalign in ["male", "trans male"]:
```
Revert changes.

j-gynn commented on scripts/screens/MediationScreen.py:
```diff
@@ -908,7 +906,7 @@ def update_search_cats(self, search_text):
 
         search_text = search_text.strip()
-        if search_text not in ("",):
+        if search_text not in [""]:
```
Revert change.

j-gynn commented on scripts/game_structure/windows.py:
```diff
@@ -32,6 +33,8 @@
     UITextBoxTweaked,
     UISurfaceImageButton,
+    UIModifiedScrollingContainer,
+    UIDropDownContainer,
```
why are these here?

j-gynn commented on scripts/events_module/short/condition_events.py:
```diff
@@ -229,7 +229,7 @@ def handle_nutrient(cat: Cat, nutrition_info: dict) -> None:
         elif MAL_PERCENTAGE >= cat_nutrition.percentage > STARV_PERCENTAGE:
             # because of the smaller 'nutrition buffer', kitten and elder should get the starving condition.
-            if cat.status in ("kitten", "elder"):
+            if cat.status.rank in [CatRank.KITTEN, CatRank.ELDER]:
```
```suggestion
            if cat.status.rank in (CatRank.KITTEN, CatRank.ELDER):
```

j-gynn commented on scripts/events_module/short/condition_events.py:
```diff
@@ -304,16 +304,16 @@ def handle_illnesses(cat, season=None):
                     "running nose",
                     "stomachache",
-                ):
+                ]:
```
Revert change.

j-gynn commented on scripts/events_module/short/condition_events.py:
```diff
@@ -380,7 +380,7 @@ def handle_injuries(cat, random_cat=None):
                 not triggered
                 and cat.personality.trait
-                in (
+                in [
```
Revert change.

j-gynn commented on scripts/events_module/short/condition_events.py:
```diff
@@ -392,7 +392,7 @@ def handle_injuries(cat, random_cat=None):
                     "vengeful",
                     "impulsive",
-                )
+                ]
```
Revert change.

j-gynn commented on scripts/events_module/short/condition_events.py:
```diff
@@ -618,7 +622,7 @@ def handle_already_ill(cat):
                 cat.illnesses.pop(illness)
                 # make sure complications get reset if infection or fester were healed
-                if illness in ("an infected wound", "a festering wound"):
+                if illness in ["an infected wound", "a festering wound"]:
```
Revert change.

j-gynn commented on scripts/events_module/short/condition_events.py:
```diff
@@ -920,14 +927,14 @@ def handle_already_disabled(cat):
                     if cat.parent1 in Cat.all_cats:
                         parent1_dead = Cat.all_cats[cat.parent1].dead
-                        if Cat.all_cats[cat.parent1].status == "medicine cat":
+                        if Cat.all_cats[cat.parent1].status.rank.is_any_medicine_rank():
```
Not like-for-like as this also includes medicine cat apprentice.

Whilst I agree this probably *should* include apprentices, with such a chomky PR I'd rather aim for exact behaviour and write issues to do this change later.

scribblecrumb replied:
I went with including apps here because med apps can (and often do?) end up adults before graduating, so I think that means they can presumably end up here?

j-gynn commented on scripts/events_module/short/condition_events.py:
```diff
@@ -920,14 +927,14 @@ def handle_already_disabled(cat):
                     if cat.parent2 in Cat.all_cats:
                         parent2_dead = Cat.all_cats[cat.parent2].dead
-                        if Cat.all_cats[cat.parent2].status == "medicine cat":
+                        if Cat.all_cats[cat.parent2].status.rank.is_any_medicine_rank():
```
see above.

j-gynn commented on scripts/events_module/short/condition_events.py:
```diff
@@ -981,17 +988,7 @@ def determine_retirement(cat, triggered):
-                "mediator apprentice",
-                "elder",
-            ]
+            and cat.status.rank in (CatRank.APPRENTICE, CatRank.WARRIOR)
```
Something within the function should ensure no kittypets and co. can pass this check. Right now, they can.

scribblecrumb replied:
Kittypets would be `CatRank.KITTYPET`.  The only cats passing this check should be clancats.  I could set it to check for `CatSocial.CLANCAT`, but that seems redundant.

j-gynn commented on scripts/cat/status.py:
```diff
@@ -0,0 +1,639 @@
+        CatRank.ROGUE: CatSocial.ROGUE,
+        CatRank.KITTYPET: CatSocial.KITTYPET,
+    }
+    """A dict with rank as the key and it's matching social status as the value"""
```
```suggestion
    """A dict of ranks and their corresponding social status"""
```

j-gynn commented on scripts/cat/status.py:
```diff
@@ -0,0 +1,639 @@
+        # if no group_history was given, we'll see if any other info was given that we can build it with
+        if not self.group_history and (rank or age):
+            self.generate_new_status(social=social, group=group, rank=rank, age=age)
+        # if we didn't get any information at all, we're gonna default to a warrior
```
We should *probably* raise a warning at least, since that's kind of bizarre behaviour.

scribblecrumb replied:
What do you mean?  This is generally how any new cat has their status first generated.  The only cats with existing group_history will be cats being loaded from a save.

j-gynn commented on scripts/cat/status.py:
```diff
@@ -0,0 +1,639 @@
+                    rank = self.get_rank_from_age(age)
+                else:  # god this should never happen, but I'm paranoid
+                    rank = CatRank.WARRIOR
+            rank = CatRank(rank)
```
This might throw an error when you pass in `CatRank.WARRIOR` from the else block above...?

j-gynn commented on scripts/cat/status.py:
```diff
@@ -0,0 +1,639 @@
+        if social and not isinstance(social, CatSocial):
+            print(f"ERROR: {social} is not a valid social")
+        if group and not isinstance(group, CatGroup):
+            print(f"ERROR: {group} is not a valid group")
```
This is an unrecoverable situation and should crash the game if encountered.  Probably `TypeError`?

j-gynn commented on scripts/cat/status.py:
```diff
@@ -0,0 +1,639 @@
+        """
+        Returns a dict of past held ranks. Key is rank, value is moons spent as that rank.
+        """
+        history = {}
```
could set this up as a default dict instead of the chaos that happens below?

j-gynn commented on scripts/cat/status.py:
```diff
@@ -0,0 +1,639 @@
+
+    def find_prior_clan_rank(self, clan: CatGroup = None):
+        """
+        Finds the last clan rank held of a current outsider
```
```suggestion
        Finds the last held clan rank of a current outsider
```

j-gynn commented on scripts/cat/status.py:
```diff
@@ -0,0 +1,639 @@
+
+        return past_ranks[-1]
+
+    def in_player_clan(self) -> bool:
```
```suggestion
    def alive_in_player_clan(self) -> bool:
```

As I mentioned elsewhere, I think this should explicitly mark that a cat is *alive* and part of the Clan. I know it's semantics, but to my mind a dead cat is still part of PlayerClan, they're just also in StarClan now. It would also make several statements more explicit, which is always good.

j-gynn replied:
These should all probably be properties too.

j-gynn commented on scripts/cat/status.py:
```diff
@@ -0,0 +1,639 @@
+        """
+        Returns True if the cat is currently part of the player clan.
+        """
+        return True if self.group == CatGroup.PLAYER_CLAN else False
```
```suggestion
        return self.group == CatGroup.PLAYER_CLAN
```

j-gynn commented on scripts/cat/status.py:
```diff
@@ -0,0 +1,639 @@
+        Returns True if the cat isn't part of a clan.
+        """
+
+        return True if self.social != CatSocial.CLANCAT else False
```
```suggestion
        return self.social != CatSocial.CLANCAT
```

j-gynn commented on scripts/cat/status.py:
```diff
@@ -0,0 +1,639 @@
+        """
+        Returns True if the cat is currently a clancat in any clan.
+        """
+        return True if self.social == CatSocial.CLANCAT else False
```
```suggestion
        return self.social == CatSocial.CLANCAT
```

j-gynn commented on scripts/cat/status.py:
```diff
@@ -0,0 +1,639 @@
+        return False
+
+    def is_exiled(self, group: CatGroup = None) -> bool:
+        """
```
mising param argument of docstring

j-gynn commented on scripts/cat/status.py:
```diff
@@ -0,0 +1,639 @@
+
+    def is_near(self, group: CatGroup) -> bool:
+        """
+        Returns True if the cat is near the specified group
```
missing params

j-gynn commented on scripts/cat/status.py:
```diff
@@ -0,0 +1,639 @@
+            print(
+                "WARNING: group history could not be made due to missing age and rank information"
+            )
+            return
```
I think this should raise an exception rather than just silent return.

j-gynn commented on scripts/clan.py:
```diff
@@ -167,7 +167,7 @@ def __init__(
 
         self.starting_members = starting_members
-        if game_mode in ("expanded", "cruel season"):
+        if game_mode in ["expanded", "cruel season"]:
```
Revert change.

j-gynn commented on scripts/clan.py:
```diff
@@ -1285,7 +1232,8 @@ def temperament(self):
             i
             for i in Cat.all_cats_list
-            if i.status not in ("leader", "deputy") and not i.dead and not i.outside
+            if i.status.rank not in [CatRank.LEADER, CatRank.DEPUTY]
```
```suggestion
            if i.status.rank not in (CatRank.LEADER, CatRank.DEPUTY)
```

j-gynn commented on scripts/clan.py:
```diff
@@ -1304,28 +1252,28 @@ def temperament(self):
                 weight
                 * statistics.mean(
-                    [i.personality.sociability for i in (leader, deputy) if i]
+                    [i.personality.sociability for i in [leader, deputy] if i]
```
Revert change.

j-gynn commented on scripts/clan.py:
```diff
@@ -1304,28 +1252,28 @@ def temperament(self):
                 weight
                 * statistics.mean(
-                    [i.personality.aggression for i in (leader, deputy) if i]
+                    [i.personality.aggression for i in [leader, deputy] if i]
```
Revert change.

j-gynn commented on scripts/clan.py:
```diff
@@ -1304,28 +1252,28 @@ def temperament(self):
             clan_sociability = round(
                 statistics.mean(
-                    [i.personality.sociability for i in (leader, deputy) if i]
+                    [i.personality.sociability for i in [leader, deputy] if i]
```
Revert change.

j-gynn commented on scripts/clan.py:
```diff
@@ -1304,28 +1252,28 @@ def temperament(self):
             clan_aggression = round(
                 statistics.mean(
-                    [i.personality.aggression for i in (leader, deputy) if i]
+                    [i.personality.aggression for i in [leader, deputy] if i]
```
Revert change.

j-gynn commented on scripts/events.py:
```diff
@@ -108,15 +96,13 @@ def one_moon(self):
 
         if (
-            game.clan.game_mode in ("expanded", "cruel season")
+            game.clan.game_mode in ["expanded", "cruel season"]
```
Revert change.

j-gynn commented on scripts/events.py:
```diff
@@ -190,11 +176,15 @@ def one_moon(self):
+                    alive_cats = list(
+                        filter(
+                            lambda kitty: (
+                                kitty.status.rank != CatRank.LEADER
```
why forbid leader here?

j-gynn commented on scripts/events.py:
```diff
@@ -248,7 +238,7 @@ def one_moon(self):
 
         if (
-            game.clan.game_mode in ("expanded", "cruel season")
+            game.clan.game_mode in ["expanded", "cruel season"]
```
Revert change.

j-gynn commented on scripts/events.py:
```diff
@@ -266,14 +256,14 @@ def one_moon(self):
         )
 
-        if game.clan.game_mode in ("expanded", "cruel season"):
+        if game.clan.game_mode in ["expanded", "cruel season"]:
```
Revert change.

j-gynn commented on scripts/events.py:
```diff
@@ -423,11 +411,9 @@ def handle_lead_den_event(self):
+                    outsider_cat.status.change_group_nearness(CatGroup.PLAYER_CLAN)
 
-                elif info_dict["interaction_type"] in ("invite", "search"):
+                elif info_dict["interaction_type"] in ["invite", "search"]:
```
Revert change.

j-gynn commented on scripts/events.py:
```diff
@@ -619,24 +599,30 @@ def handle_focus(self):
-            ]
+            healthy_warriors = list(
+                filter(
+                    lambda c: c.status.rank.is_any_adult_warrior_like_rank()
```
does this exclude adolescent warriors? /genq

scribblecrumb replied:
It would not, which might be an issue with the name hmmm

I went with `is_any_adult_warrior_like_rank` because I wanted it to exclude the apps. I thought not having `adult` in there would be unclear, but yeah it's still an issue with adolescent warriors.  Maybe `is_any_graduated_warrior_rank`? idkkkk

j-gynn commented on scripts/events.py:
```diff
@@ -725,9 +718,9 @@ def handle_focus(self):
-                    if status_use in ("deputy", "leader"):
-                        status_use = "warrior"
+                    status_use = cat.status.rank
+                    if status_use in [CatRank.DEPUTY, CatRank.LEADER]:
```
```suggestion
                    if status_use in (CatRank.DEPUTY, CatRank.LEADER):
```

j-gynn commented on scripts/events.py:
```diff
@@ -1228,47 +1192,50 @@ def perform_ceremonies(self, cat):
             if (
                 not cat.no_retire
-                and cat.status in ["warrior", "deputy"]
+                and cat.status.rank in [CatRank.WARRIOR, CatRank.DEPUTY]
```
```suggestion
                and cat.status.rank in (CatRank.WARRIOR, CatRank.DEPUTY)
```

j-gynn commented on scripts/events.py:
```diff
@@ -1625,15 +1588,15 @@ def ceremony(self, cat, promoted_to, preparedness="prepared"):
         # getting the random honor if it's needed
         random_honor = None
-        if promoted_to in ["warrior", "mediator", "medicine cat"]:
+        if promoted_to in [CatRank.WARRIOR, CatRank.MEDIATOR, CatRank.MEDICINE_CAT]:
```
```suggestion
        if promoted_to in (CatRank.WARRIOR, CatRank.MEDIATOR, CatRank.MEDICINE_CAT):
```

j-gynn commented on scripts/events.py:
```diff
@@ -1625,15 +1588,15 @@ def ceremony(self, cat, promoted_to, preparedness="prepared"):
                 random_honor = i18n.t("defaults.ceremony_honor")
 
-        if cat.status in ["warrior", "medicine cat", "mediator"]:
+        if cat.status.rank in [CatRank.WARRIOR, CatRank.MEDICINE_CAT, CatRank.MEDIATOR]:
```
```suggestion
        if cat.status.rank in (CatRank.WARRIOR, CatRank.MEDICINE_CAT, CatRank.MEDIATOR):
```

j-gynn reviewed (changes requested):
okay......... it is done....... 139 comments. bon appetit.

scribblecrumb committed:
cats.py fixes

scribblecrumb committed:
starclan enum name change

scribblecrumb committed:
skills and names fixes

scribblecrumb committed:
utility fixes

scribblecrumb committed:
myriad of fixes

scribblecrumb committed:
fixes to status class!

scribblecrumb committed:
formatting for former clancats

scribblecrumb committed:
fix the patrol merge

scribblecrumb committed:
patrol fixes

scribblecrumb committed:
preg fixes

scribblecrumb committed:
rename in_player_clan

scribblecrumb committed:
rename in_player_clan

scribblecrumb committed:
short event merge fixes

scribblecrumb committed:
is_leader()

scribblecrumb committed:
screen fixes

scribblecrumb committed:
status fixes

includes switch some funcs to properties

scribblecrumb committed:
final fixes

scribblecrumb committed:
Merge remote-tracking branch 'upstream/development' into status-update

scribblecrumb committed:
add a little check for none

scribblecrumb committed:
reformatting

scribblecrumb committed:
removing status check

just yeeting this line.  this is the fallback anyways, we should really never be ending up with these strings.  eventually this is smth to fix by just uniting our death strings instead of having a lead format and a non-lead format.

scribblecrumb commented:
@j-gynn bon appetit

j-gynn commented on scripts/cat/cats.py:
```diff
@@ -923,16 +923,15 @@ def add_to_clan(self) -> list:
         old_rank = self.status.rank
 
-        self.status.change_rank(new_rank)
+        self.status._change_rank(new_rank)
```
may want to add the pylint/pycharm inline ignores to this, or at the very least add a comment saying "yes this is private, in this one instance it's fine"

j-gynn commented on scripts/cat/cats.py:
```diff
@@ -1528,10 +1523,10 @@ def one_moon(self):
         self.moons += 1
         if self.moons == 1 and self.status.rank == CatRank.NEWBORN:
-            self.status.change_rank(CatRank.KITTEN)
+            self.status._change_rank(CatRank.KITTEN)
```
shouldn't this use the cat version `rank_change` /genq

scribblecrumb replied:
It doesn't need to tbh, `cat.rank_change()` does a lot of things, but newborns don't need them.  They just need to change their rank, they have no mentors to update or lists to be added to.  Generally `status._change_rank()` isn't what we want bc most ranks have other related attributes to update, but newborns are simple.

j-gynn commented on scripts/cat/cats.py:
```diff
@@ -2185,7 +2180,11 @@ def is_disabled(self):
     def available_to_work(self):
-        return not self.dead and self.status.in_player_clan() and not self.not_working()
+        return (
+            not self.dead
```
doesn't `not self.dead` get handled by `self.status.alive_in_player_clan`?

j-gynn commented on scripts/cat/status.py:
```diff
@@ -92,30 +82,32 @@ def __init__(
+            if social and not isinstance(social, CatSocial):
+                print(f"ERROR: {social} is not a valid social")
+            if group and not isinstance(group, CatGroup):
+                print(f"ERROR: {group} is not a valid group")
```
```suggestion
        if rank and not isinstance(rank, CatRank):
            if rank.casefold() in ("exiled", "lost", "former clancat"):
                if age:
                    rank = self.get_rank_from_age(age)
                else:  # god this should never happen, but I'm paranoid
                    rank = CatRank.WARRIOR
            rank = CatRank(rank)
        if social and not isinstance(social, CatSocial):
            if social.casefold() == "former clancat":
                social = CatSocial.CLANCAT
            social = CatSocial(social)
        if group and not isinstance(group, CatGroup):
            group = CatGroup(group)

        if rank and not isinstance(rank, CatRank):
            raise TypeError("{rank} is not a valid rank")
        if social and not isinstance(social, CatSocial):
            raise TypeError("{social} is not a valid rank")
        if group and not isinstance(group, CatGroup):
            raise TypeError("{group} is not a valid rank")
```

j-gynn commented on scripts/cat/status.py:
```diff
@@ -175,10 +167,9 @@ def _start_group_history(
 
         if not age and not rank:
-            print(
+            raise Exception(
```
A more precise exception would be better here.
```suggestion
            raise ValueError(
```

j-gynn commented on scripts/cat/status.py:
```diff
@@ -435,16 +449,17 @@ def add_to_group(
             new_rank = self.find_prior_clan_rank()
             # we don't need to change leaders and deps if they're going to an afterlife
             if (
                 new_rank in [CatRank.LEADER, CatRank.DEPUTY]
```
```suggestion
                new_rank in (CatRank.LEADER, CatRank.DEPUTY)
```

j-gynn commented on scripts/cat/status.py:
```diff
@@ -503,27 +518,23 @@ def change_group_nearness(self, group: CatGroup):
-                else:
-                    entry["near"] = True
+                if entry["near"]:
+                    entry["near"] = not entry.get("near", False)
```
should just be able to say `not entry.get("near")`, since `None` is falsy iirc. Even so, I almost want to say `not entry["near"]` so that it intentionally crashes if no near data exists, since that's a sign something's wrong

j-gynn commented on scripts/events_module/outsider_events.py:
```diff
@@ -40,7 +40,7 @@ def killing_outsiders(cat: Cat):
                 else:
                     text = (
-                        f"Rumors reach your Clan that the {cat.status.social} "
+                        f"Rumors reach your Clan that the {cat.status.social.replace('_', '')} "
```
meant to be a space in here? /genq
(also ugh, hardcoded).

j-gynn commented on scripts/screens/RoleScreen.py:
```diff
@@ -330,12 +330,12 @@ def update_disabled_buttons(self):
 
         if game.clan.leader:
-            leader_invalid = game.clan.leader.status.in_player_clan()
+            leader_invalid = game.clan.leader.status.alive_in_player_clan
```
shouldn't this be `not [...]` ? Since it sets the button for "promote leader" if leader_invalid is true. This will return true if the leader is currently in the clan (i.e., valid).

j-gynn commented on scripts/screens/RoleScreen.py:
```diff
@@ -330,12 +330,12 @@ def update_disabled_buttons(self):
 
         if game.clan.deputy:
-            deputy_invalid = game.clan.deputy.status.in_player_clan()
+            deputy_invalid = game.clan.deputy.status.alive_in_player_clan
```
same as above.

j-gynn commented on scripts/utility.py:
```diff
@@ -583,7 +583,7 @@ def create_new_cat_block(
                 chosen_cat.add_to_clan()
                 if chosen_cat.status.rank != rank:
-                    chosen_cat.status.change_rank(rank)
+                    chosen_cat.status._change_rank(rank)
```
shouldn't this be `chosen_cat.rank_change`? /genq

j-gynn commented on scripts/utility.py:
```diff
@@ -825,10 +826,8 @@ def create_new_cat(
-        elif moons >= 12:
+        else:
             rank = CatRank.WARRIOR
-        elif moons >= 120:
```
I think we should still have this check, just put it before `CatRank.WARRIOR` as `elif moons >= 120`, leaving the range `12-119` for warrior.

scribblecrumb committed:
private use comment

scribblecrumb committed:
remove unnecessary check

scribblecrumb committed:
switch to raising typeerrors

scribblecrumb committed:
fixes to status.py

scribblecrumb committed:
missing space

scribblecrumb committed:
fix rolescreen invalids

scribblecrumb committed:
new cat creation suggestions

scribblecrumb committed:
Merge remote-tracking branch 'upstream/development' into status-update

scribblecrumb committed:
formatter

j-gynn commented on scripts/utility.py:
```diff
@@ -1455,15 +1457,6 @@ def gather_cat_objects(
-            clan_cats = [x for x in Cat.all_cats_list if x.status.alive_in_player_clan]
-            out_set.update(
-                sample(clan_cats, randint(1, max(1, round(len(clan_cats) / 8))))
-            )
```
intended removal...?

scribblecrumb replied:
Yup intended! If you scroll down you'll see that those two abbrs just got moved to after patrol and multi are checked

j-gynn reviewed (commented):
Just this one removal that might be a mistake. Otherwise... I think we're done?
